tests/votinghistory.py

```diff
@@ -0,0 +1,164 @@
+            )
+
+        # create json file to be passed as the argument to the query.lua file
+        # it is passing a member id
+        with open("param.json", "w") as pfile:
+            pfile.write("""{"p": 0}""")
+
+        # propose to add a new member
+        # proposal number 0
+        infra.proc.ccall("./genesisgenerator", "cert", "--name=member4")
+        result = infra.proc.ccall(
+            "./memberclient",
+            "add_member",
+            "--cert=member1_cert.pem",
+            "--privk=member1_privk.pem",
+            "--host={}".format(primary.host),
+            "--port={}".format(primary.tls_port),
+            "--member_cert=member4_cert.pem",
+            "--ca=networkcert.pem",
+        )
+
+        # when proposal is added the proposal id and the result of running complete proposal are returned
+        j_result = json.loads(result.stdout)
+        assert not j_result["result"]["completed"]
+        assert j_result["result"]["id"] == 0
+
+        # 2 out of 3 members vote to accept the new member so that that member can send its own proposals
+        result = infra.proc.ccall(
+            "./memberclient",
+            "vote",
+            "--accept",
+            "--cert=member1_cert.pem",
+            "--privk=member1_privk.pem",
+            "--host={}".format(primary.host),
+            "--port={}".format(primary.tls_port),
+            "--id=0",
+            "--ca=networkcert.pem",
+            "--sign",
+        )
+        j_result = json.loads(result.stdout)
+        assert not j_result["result"]
+
+        result = infra.proc.ccall(
+            "./memberclient",
+            "vote",
+            "--accept",
+            "--cert=member2_cert.pem",
+            "--privk=member2_privk.pem",
+            "--host={}".format(primary.host),
+            "--port={}".format(primary.tls_port),
+            "--id=0",
+            "--ca=networkcert.pem",
+            "--sign",
+        )
+        j_result = json.loads(result.stdout)
+        assert j_result["result"]
+
+        # member 4 ack
+        result = infra.proc.ccall(
+            "./memberclient",
+            "ack",
+            "--cert=member4_cert.pem",
+            "--privk=member4_privk.pem",
+            "--host={}".format(primary.host),
+            "--port={}".format(primary.tls_port),
+            "--ca=networkcert.pem",
+        )
+        j_result = json.loads(result.stdout)
+        assert j_result["result"]
+
+        ledger_filename = network.find_leader()[0].remote.get_ledger_full_path()
+
+    l = ledger.Ledger(ledger_filename)
+
+    # this maps a member_id to a cert object, and is updated when we iterate the transactions,
+    # so that we always have the correct cert for a member on a given transaction
+    members = {}
+    verified_votes = 0
+    for tr in l:
+        tables = tr.get_public_domain().get_tables()
+        members_table = tables["membercerts"]
+        for cert, member_id in members_table.items():
```

Comment:
Should we only do this after the condition on line `139`?

Reply:
Member certs may be added on earlier transactions than the ones that contain the voting, so we need to store them in order to be able to query at a later stage.